src/store/Controller.cc

```diff
@@ -855,49 +858,51 @@ Store::Controller::syncCollapsed(const sfileno xitIndex)
     collapsed->setCollapsingRequirement(true);
 }
 
-/// Called for Transients entries that are not yet anchored to a cache.
+/// If possible and has not been done, associates the entry with its store(s).
 /// \returns false for not-yet-cached entries that we may attach later
 /// \returns true for other entries after synchronizing them with their store
-/// and setting inSync to reflect that synchronization outcome.
 bool
-Store::Controller::anchorToCache(StoreEntry &entry, bool &inSync)
+Store::Controller::anchorToCache(StoreEntry &entry)
 {
     assert(entry.hasTransients());
     assert(transientsReader(entry));
 
+    // TODO: Attach entries to both memory and disk
+
+    // TODO: Reduce code duplication with syncCollapsed()
+    if (sharedMemStore && entry.mem().memCache.io == MemObject::ioDone) {
+        debugs(20, 5, "already handled by memory store: " << entry);
+        return true;
+    } else if (sharedMemStore && entry.hasMemStore()) {
```

Comment:
Is it possible that `entry.mem().memCache.io == MemObject::ioDone` but  entry.hasMemStore() is true - i.e., can we join these two conditions?

Reply:
> Is it possible that `entry.mem().memCache.io == MemObject::ioDone` but `entry.hasMemStore()` is true

I do not think it is possible today, but it might become common as we continue to separate shared memory cache from the worker-specific memory storage. AFAICT, there is no code ensuring this invariant today -- MemStore just marks entries with ioDone when it refuses to cache them OR is done caching them, releasing the lock.

> can we join these two conditions?

We can just check for the `!(false and false)` condition, but I doubt it will improve this code; it will just make the same information a bit more cryptic for the cache.log (and code) reader.

| io | hasMemStore() | meaning |
| --- | -------------------- | ----------- |
| 0 | 0 | has not attempted to mem-cache yet |
| 0 | 1 | is attempting to mem-cache now |
| 1 | 0 | attempted to mem-cache and not holding the lock |
| 1 | 1 | attempted to mem-cache and still holding the lock |

Also, any changes here would increase the difference with syncCollapsed() unless we modify that method too. Perhaps that can be done later, when syncCollapsed() is polished to handle exceptions and the code duplication between the two methods is removed/reduced.

Reply:
Ok.
Don't we need a comment explaining why we have the "already handled by memory store..." condition for Memory store but don't have such condition for Disk store?

Reply:
We could try to explain the current shared memory store deficiencies here (and in a few other places, I am sure!), but I would rather not do that given the complexity of that explanation and the fact that we are actively working on _removing_ those deficiencies. When we are done, the (shared) memory store condition will be the same as the disk store condition. The shared memory store will no longer use the current copyFromShm() hack, eliminating the need for the current ioDone check.